feat(layout): Add support for responsive layout

obartra opened the pull request:
- Use provider to configure breakpoints through `displayAliases`

Closes https://github.com/obartra/reflex/issues/251

arahansen commented on src/withResolution.js:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    componentWillReceiveProps({ show }: Props) {
+      this.addMediaQueryListener(show)
```
this will remove and re-add the listener on every render?

obartra replied:
Good call, I can make it run only when `show` changes

arahansen commented on src/withResolution.js:
```diff
@@ -0,0 +1,86 @@
+
+type State = { shouldShow: boolean }
+
+export default function withResolution(Component: *) {
```
Just a heads up: the usage of web-specific APIs will prove problematic for React Native (`window`,

obartra replied:
do we have a way to detect screen resolution changes that would work well with RN?

arahansen replied:
There are built in APIs to RN that detect resolution. They're not compatible with web APIs though. 

Probably fine to go this route for now. And punt on resolution detection, since we don't even have a native solution in place 😛 .

I think the use case for a `show` api is rarer for native. Maybe something more focused on changing `size` attributes based on screen size. the `show` api can accomplish that, but it'll be a bit verbose.

obartra replied:
Ok for now I added a step to skip `withResolution` if `window.matchMedia` is not available and log a warning. As you suggest, we can worry about a proper solution when we actually support RN.

I imagine we may want different layouts in landscape vs portrait mode or something like that but probably not much more beyond that

obartra commented:
94bb07e changes from tracking whether to mount/unmount the hidden components to setting `display: none` when it needs to be hidden.

obartra commented:
Changed approach based on discussion with @arahansen, now I added a `mediaQuery.js` to manage it. This ensures there are at most as many eventListeners are `displayAliases` keys.

It now generates an object in `withResolution` that looks like `{ small: true, large: false}` so that `Base` *could* consume it easily if we wanted to change the API for `margin` / `padding` / `size` to support a format like: `size={{ small: 6, large: 12 }}`

obartra commented on package.json:
```diff
@@ -114,7 +114,7 @@
     {
       "path": "./dist/reflex.min.js",
-      "maxSize": "8 Kb"
+      "maxSize": "9.5 Kb"
```
This feature is fairly large as it is now (1.5kb more), I'm open to removing features if we think it's worth it

obartra commented on src/reflex.js:
```diff
@@ -1,30 +1,38 @@
+  Size,
   SpacingProps,
+  SpacingValues,
 } from './types'
```
I grouped the exports and moved `GridProps` to types now that it's gotten a bit more complex

obartra commented on src/types.js:
```diff
@@ -59,9 +63,70 @@ export type SpacingAliases = {
+  show?: DisplayValues,
+  size?: Size,
+  style?: { [string]: string | number },
+}
```
`OneResolutionGrid` is used internally only, after `withResolution` has resolved the values to use

andrewmtoy replied:
If this is not reused, can we move it to the file where it is being used?

obartra commented on src/types.js:
```diff
@@ -59,9 +63,70 @@ export type SpacingAliases = {
+  'paddingRight',
+  'paddingTop',
+  'size',
+]
```
These are the properties for which you can now pass either an object (with keys indicating when to apply it) or the previous format

obartra commented on src/utils.js:
```diff
@@ -40,6 +44,42 @@ export function validateSpacingProps(props: SpacingProps) {
+  if (to) {
+    response.push(`(max-width: ${to}px)`)
+  }
+  return response.join(' and ')
```
This is one of the limitations of this approach. Right now the only media queries handled are `min/max width`)

obartra commented on src/withResolution/mediaQuery.js:
```diff
@@ -0,0 +1,63 @@
+    +alias: string,
+    +callback: Callback,
+  |}>,
+}
```
media queries manages setting up the minimal number of listeners possible and removing them when they are no longer in use, based on some quick jsfiddles this roughly doubles performance on Chrome

obartra commented on storybook/stories/grid/resolution.js:
```diff
@@ -0,0 +1,68 @@
+      padding={2}
+      margin={[1, 0]}
+      show={['small', 'large']}
+      size={{ small: 12, large: 6 }}
```
I updated the stories to contain both the `show` property and the new `object` format (for some `size` props)

arahansen commented on src/defaults.json:
```diff
@@ -1,4 +1,9 @@
 {
   "spacingAliases": {},
+  "displayAliases": {
+    "small": "-599",
```
It's nice how terse this syntax is, but it's a bit weird to read. Would you still prefer this to something like:
`{ small: { minWidth: 599 }, medium: { minWidth: 600, maxWidth: 899 }, ... }`

This would also be a little bit closer to how media queries are typically written (with the `minWidth/maxWidth` keywords)

obartra replied:
I guess it's a trade off between complexity of the API and flexibility but I like it because this approach enables us to add more formats later on (`e.g. { large: { orientation: 'landscape' } }`

arahansen commented on src/types.js:
```diff
@@ -59,9 +63,70 @@ export type SpacingAliases = {
+  style?: { [string]: string | number },
+}
+
+export const resolutionProperties = [
```
This is conflating flow types, and properties used by reflex. They are different thing in my mind, and maybe shouldn't be grouped together in the same module.

obartra replied:
that's fair, I was thinking about this, imo it made sense to put them here so that there's a single spot to look at when new properties are added. Maybe the right solution would be to generate this array but that also feels overkill

andrewmtoy replied:
I also find this confusing.

arahansen commented on src/types.js:
```diff
@@ -59,9 +63,70 @@ export type SpacingAliases = {
+  size?: Size | { [string]: Size },
+}
+
+export default undefined
```
I dont think you need to make eslint happy with this since you have multiple exports.

obartra replied:
🙈

arahansen commented on src/withResolution/mediaQuery.js:
```diff
@@ -0,0 +1,63 @@
+  }
+}
+
+export const supportsMatchMedia = !!window.matchMedia
```
Could this be more robust? something like `window && typeof window.matchMedia === 'function'`

Mostly just thinking about the React Native use case where `window` will not be available.

Although the more I think about it, the more I think that RN will need its own implementation of `withResolution` since most of the logic happening in this module will not really work in a RN environment

obartra replied:
it's still a good point, we shouldn't assume window, will update

arahansen commented on src/withResolution/mediaQuery.js:
```diff
@@ -0,0 +1,63 @@
+const perCallbackData: PerCallbackData = {}
+const queryData: QueryData = {}
+
+function callCallbacks(query) {
```
Maybe this could be called something like `getCallCallbacks` since this is a HOF whose returned function is the one that calls callbacks?

andrewmtoy replied:
What about `getCallbacksInvoker`?

arahansen commented on src/withResolution/mediaQuery.spec.js:
```diff
@@ -0,0 +1,79 @@
+import { register, unregister } from './mediaQuery'
+
+describe('Media Query Manager', () => {
```
would we be able to do `spyOn(global.matchMedia)` (and/or `addListener`)  to validate that we are only adding as many listeners as we actually need?

obartra replied:
oh interesting, will do

obartra replied:
thinking about it more, we cannot attach a listener directly because `matchMedia` creates a new object to which you can attach queries. That's the part that's already tested

andrewmtoy commented on test/env.js:
```diff
@@ -4,6 +4,29 @@ global.requestAnimationFrame = jest.fn().mockImplementation(() => {
+        i -= 1
+      }
+    },
+    onchange: () => allListeners[media].forEach(listener => listener({})),
```
nit: onChange

obartra replied:
this is part of the MediaQueryList spec https://drafts.csswg.org/cssom-view/#dom-mediaquerylist-onchange

andrewmtoy commented on storybook/stories/grid/spacingAliases.md:
```diff
@@ -34,5 +34,6 @@ Changing base to `24` changes the spacing values:
+Use the knobs to try different `base` values.
 
-Use the knobs to try different `base` values. For more information on how to use the `ConfigProvider` check the [wiki page](https://github.com/obartra/reflex/wiki/ConfigProvider).
+There are other parameters that can be customized through `ConfigProvider`. For more information check the [wiki page](https://github.com/obartra/reflex/wiki/ConfigProvider).
```
I think having a hide prop in addition to show would be helpful.

andrewmtoy commented on storybook/stories/grid/resolution.md:
```diff
@@ -0,0 +1,21 @@
+`show` allows you to define when a given component is visible based on the screen width.
```
I think having a hide prop in addition to show would be helpful.

obartra replied:
That's a good idea. I think as far as we don't allow having `show` and `hide` set at the same time there should be no issues. I'll take a look if I don't add it here I may create a ticket for a follow up PR

obartra replied:
I've added https://github.com/obartra/reflex/issues/307

andrewmtoy commented on src/withResolution/mediaQuery.js:
```diff
@@ -0,0 +1,63 @@
+// @flow
+type Callback = (mql: MediaQueryList, alias: string) => void
+type QueryData = {
+  [query: string]: {|
```
You could replace this with
```
ReadOnly<{
  mediaQuery: MediaQueryList,
  allCallbacks: (mql: MediaQueryListEvent) => void,
}>
```

obartra replied:
the values themselves are read only once set but the array itself is not read only

andrewmtoy commented on src/withResolution/mediaQuery.js:
```diff
@@ -0,0 +1,63 @@
+  |},
+}
+type PerCallbackData = {
+  [query: string]: Array<{|
```
You could replace this with
```
ReadOnlyArray<{|
       alias: string,
       callback: Callback,
    |}>,
```

obartra replied:
same thing here

andrewmtoy commented on src/types.js:
```diff
@@ -59,9 +63,70 @@ export type SpacingAliases = {
+  align?: AlignGrid | { [string]: AlignGrid },
+  justify?: Justify | { [string]: Justify },
+  margin?: Spacing | { [string]: Spacing },
+  marginBottom?: SpacingValues | { [string]: SpacingValues },
```
I don't think these flowtypes are correct. https://flow.org/en/docs/types/objects/#toc-objects-as-maps

`{[string]: SpacingValues}` tells me that I can pass in this to marginBottom:
```
marginBottom={foo: 1}
```

arahansen replied:
I had noticed this too. The issue is that the possible keys are dynamic based on aliases. It's possible we can address this with generic typing. But I'm ok with a little less specific typing and addressing in a future PR.

andrewmtoy replied:
Not sure I follow.  With aliases this could be something like `marginBottom?: string | $Keys<typeof spacingAliases>` -> `marginBottom="XL"`

obartra replied:
With it you can now use:

```
<Grid marginBottom={{ large: 'M', medium 'S', small: 'XS' }} />
```

In addition to being able to say:

```
<Grid show="large" marginBottom="M" />
<Grid show="medium" marginBottom="S" />
<Grid show="small" marginBottom="XS" />
```

The difference is that with the first approach you maintain the same component so state is preserved. It won't be always possible to use the first approach which is why the second one is there.

The reason why the types are defined as `[string]` instead of `"large" | "medium" | "small"` is because these are just the defaults
